service/listeners.go

```diff
@@ -128,9 +128,9 @@ type readRequest struct {
 
 type virtualPacketConn struct {
 	net.PacketConn
-	mu          sync.Mutex // Mutex to protect access to the channels
-	readCh      chan readRequest
-	onCloseFunc OnCloseFunc
+	mu      sync.Mutex // Mutex to protect access to the channels
+	readCh  chan readRequest
+	closeCh chan struct{}
 }
 
 func (pc *virtualPacketConn) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
```

Comment:
Where is the deadlock happening? Is it on line 144?

```go
	pc.readCh <- readRequest{
		buffer: p,
		respCh: respCh,
	}
```

You could potentially use a select statement on the read and have a case with the `closeCh` if you want to return on either of those.

Or the read on `respCh` could be in a select with reading from the `closeCh`, so it returns immediately on close.

Reply:
BTW, don't you still have the problem with your change? A Close on ReadFrom won't immediately cancel the read.